mlflow/llama_index/tracer.py

```diff
@@ -293,3 +340,99 @@ def _extract_token_usage(
+    A class is responsible for closing the pending streaming spans that are waiting
+    for the stream to be exhausted. Once the associated stream is exhausted, this
+    class will resolve the span, as well as recursively resolve the parent spans
+    that returns the same (or derived) stream.
+    """
+
+    def __init__(self, client: mlflow.MlflowClient):
+        self._client = client
+        self._span_id_to_span_and_gen: Dict[str, Tuple[LiveSpan, Generator]] = {}
+
+    def is_streaming_result(self, result: Any) -> bool:
+        return (
+            inspect.isgenerator(result)  # noqa: SIM101
+            or isinstance(result, (StreamingResponse, AsyncStreamingResponse))
+            or isinstance(result, StreamingAgentChatResponse)
+            or (isinstance(result, TaskStepOutput) and self.is_streaming_result(result.output))
+        )
+
+    def register_stream_span(self, span: LiveSpan, result: Any) -> bool:
+        """
+        Register the pending streaming span with the associated generator.
+
+        Args:
+            span: The span that has a streaming output.
+            result: The streaming result that is being processed.
+
+        Returns:
+            True if the span is registered successfully, False otherwise.
+        """
+        if inspect.isgenerator(result):
+            stream = result
+        elif isinstance(result, (StreamingResponse, AsyncStreamingResponse)):
+            stream = result.response_gen
+        elif isinstance(result, StreamingAgentChatResponse):
+            stream = result.chat_stream
+        elif isinstance(result, TaskStepOutput):
+            stream = result.output.chat_stream
+        else:
+            _logger.warning(
+                f"Unsupported streaming response type: {type(result)}. The span will "
+                "be ended immediately without waiting for the stream to finish."
+            )
+            return False
+
+        if inspect.getgeneratorstate(stream) == inspect.GEN_CLOSED:
+            # Not registering the span because the generator is already exhausted
+            return False
+
+        self._span_id_to_span_and_gen[span.span_id] = (span, stream)
+        return True
+
+    def resolve(self, span: LiveSpan, event: _StreamEndEvent):
+        """
+        Finish the streaming span and recursively resolve the parent spans that
+        returns the same (or derived) stream.
+        """
+        _, stream = self._span_id_to_span_and_gen.pop(span.span_id, (None, None))
+        if not stream:
+            return
+
+        if isinstance(event, (LLMChatEndEvent, LLMCompletionEndEvent)):
+            outputs = event.response
+            status = SpanStatusCode.OK
+        elif isinstance(event, ExceptionEvent):
+            outputs = None
+            status = SpanStatusCode.ERROR
+            span.add_event(SpanEvent.from_exception(event.exception))
+        else:
+            raise ValueError(f"Unsupported event type to resolve streaming: {type(event)}")
```

Comment:
Curious on your thoughts about not raising here. If a user has a new implementation of a feature that we haven't fully covered by type checking the base classes defined here for the event types, could we terminate the span while adding a custom event message that informs the user that their trace was not gracefully resolved while not raising in the main call stack? 
Not a blocker - as this guard shouldn't be hit except in a really odd scenario - but just want to hear your opinion on this.

Reply:
For this particular line, the event handler should only be called for the fixed set of events that is registered with `@_handle_event.register`. So this is a guardrail for implementation error rather than handling unknown events.

In general, I agree that terminating span always is ideal way, but it's a bit tricky given that we don't know when to do that... There are many other events triggered, potentially including user-defined one, and we should only terminate span if it represents "streaming end".

Reply:
This makes perfect sense as this is more of a code injection behavior for interfacing with their running system. 
With the alternative here being a pre-emptive termination of the span (and a resulting loss of data for troubleshooting), raising makes sense to alert a user that **something is very wrong**. 
+1 on this approach. Thanks for talking through it! :)